File: backend/api/src/matching/trackMatcher.ts
import type { ProviderTrack, TrackMatch } from "../domain/music.js";

export function matchTracks(sourceTracks: ProviderTrack[], targetTracks: ProviderTrack[]): TrackMatch[] {
  return sourceTracks.map((source) => {
    const byIsrc = source.isrc
      ? targetTracks.find((target) => target.isrc && target.isrc === source.isrc)
      : undefined;

    if (byIsrc) {
      return {
        source,
        target: byIsrc,
        confidence: 1,
        reason: "isrc"
      };
    }

    const byTitleArtistAlbum = targetTracks.find((target) =>
      normalize(target.title) === normalize(source.title) &&
      normalize(target.artistName) === normalize(source.artistName) &&
      normalize(target.albumName ?? "") === normalize(source.albumName ?? "")
    );

    if (byTitleArtistAlbum) {
      return {
        source,
        target: byTitleArtistAlbum,
        confidence: 0.9,
        reason: "title_artist_album"
      };
    }

    const byTitleArtistDuration = targetTracks.find((target) =>
      normalize(target.title) === normalize(source.title) &&
      normalize(target.artistName) === normalize(source.artistName) &&
      isDurationClose(source.durationMs, target.durationMs)
    );

    if (byTitleArtistDuration) {
      return {
        source,
        target: byTitleArtistDuration,
        confidence: 0.78,
        reason: "title_artist_duration"
      };
    }

    return {
      source,
      confidence: 0,
      reason: "unavailable"
    };
  });
}

function normalize(value: string): string {
  return value
    .toLocaleLowerCase("en-US")
    .replace(/\s+\([^)]*\)/g, "")
    .replace(/\s+-\s+(remaster(ed)?|live|radio edit).*$/i, "")
    .replace(/\s+/g, " ")
    .trim();
}

function isDurationClose(source?: number, target?: number): boolean {
  if (!source || !target) {
    return false;
  }

  return Math.abs(source - target) <= 4_000;
}

File: backend/api/app/matching/track_matcher.py
import re
import unicodedata
from dataclasses import dataclass
from enum import Enum


class MatchReason(str, Enum):
    ISRC_EXACT = "isrc_exact"
    TITLE_ARTIST_ALBUM = "title_artist_album"
    TITLE_ARTIST_DURATION = "title_artist_duration"
    NORMALIZED_VERSION_TITLE = "normalized_version_title"
    USER_REVIEW_REQUIRED = "user_review_required"


@dataclass(frozen=True)
class MatchableTrack:
    provider: str
    provider_track_id: str
    title: str
    artist_name: str
    album_name: str | None = None
    duration_ms: int | None = None
    isrc: str | None = None


@dataclass(frozen=True)
class TrackMatchResult:
    target_track: MatchableTrack | None
    confidence: float
    reason: MatchReason


def match_track(source: MatchableTrack, candidates: list[MatchableTrack]) -> TrackMatchResult:
    exact_isrc = _find(candidates, lambda track: _same_isrc(source, track))
    if exact_isrc:
        return TrackMatchResult(exact_isrc, 1.0, MatchReason.ISRC_EXACT)

    title_artist_album = _find(
        candidates,
        lambda track: _same_title_artist(source, track) and _same_album(source, track),
    )
    if title_artist_album:
        return TrackMatchResult(title_artist_album, 0.92, MatchReason.TITLE_ARTIST_ALBUM)

    title_artist_duration = _find(
        candidates,
        lambda track: _same_title_artist(source, track) and _duration_close(source, track),
    )
    if title_artist_duration:
        return TrackMatchResult(title_artist_duration, 0.84, MatchReason.TITLE_ARTIST_DURATION)

    normalized_version_title = _find(
        candidates,
        lambda track: _normalize_version_title(source.title) == _normalize_version_title(track.title)
        and _normalize_text(source.artist_name) == _normalize_text(track.artist_name),
    )
    if normalized_version_title:
        return TrackMatchResult(
            normalized_version_title,
            0.7,
            MatchReason.NORMALIZED_VERSION_TITLE,
        )

    return TrackMatchResult(None, 0.0, MatchReason.USER_REVIEW_REQUIRED)


def _find(candidates: list[MatchableTrack], predicate) -> MatchableTrack | None:
    return next((candidate for candidate in candidates if predicate(candidate)), None)


def _same_isrc(source: MatchableTrack, candidate: MatchableTrack) -> bool:
    return bool(source.isrc and candidate.isrc and source.isrc.upper() == candidate.isrc.upper())


def _same_title_artist(source: MatchableTrack, candidate: MatchableTrack) -> bool:
    return (
        _normalize_text(source.title) == _normalize_text(candidate.title)
        and _normalize_text(source.artist_name) == _normalize_text(candidate.artist_name)
    )


def _same_album(source: MatchableTrack, candidate: MatchableTrack) -> bool:
    return bool(
        source.album_name
        and candidate.album_name
        and _normalize_text(source.album_name) == _normalize_text(candidate.album_name)
    )


def _duration_close(source: MatchableTrack, candidate: MatchableTrack, tolerance_ms: int = 3000) -> bool:
    if source.duration_ms is None or candidate.duration_ms is None:
        return False
    return abs(source.duration_ms - candidate.duration_ms) <= tolerance_ms


def _normalize_text(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value).encode("ascii", "ignore").decode("ascii")
    normalized = normalized.lower()
    normalized = re.sub(r"&", " and ", normalized)
    normalized = re.sub(r"[^a-z0-9]+", " ", normalized)
    return re.sub(r"\s+", " ", normalized).strip()


def _normalize_version_title(value: str) -> str:
    normalized = _normalize_text(value)
    normalized = re.sub(
        r"\b(remaster(?:ed)?|mono|stereo|explicit|clean|deluxe|bonus track|radio edit|edit)\b",
        " ",
        normalized,
    )
    normalized = re.sub(r"\b(19|20)\d{2}\b", " ", normalized)
    return re.sub(r"\s+", " ", normalized).strip()
